www/widgets/rt.js

```diff
@@ -291,6 +291,74 @@ static tkinter(options) {
     return {Tk: wrap_class(Tk), Canvas: wrap_class(Canvas), font: { Font: wrap_class(Font) }}
 }
 
+static dukpy() {
+    if (!crossOriginIsolated) {
+        console.error("No cross-origin isolation; dukpy will not be available.");
+    }
+
+    class JSRuntimeError {
+        constructor(msg) {
+            this.msg = msg;
+        }
+    }
+    
+    class JSInterpreter {
+        constructor() {
+            this.function_table = {};
+            this.worker = new Worker("/widgets/dukpy.js");
+            this.worker.onmessage = this.onmessage.bind(this);
+            this.buffer = new SharedArrayBuffer(1024);
+            this.write_buffer = new Int32Array(this.buffer, 4);
+            this.flag_buffer = new Int32Array(this.buffer, 0, 1);
+            this.promise_stack = [];
+            this.worker.postMessage({
+                "type": "array",
+                "buffer": this.buffer,
+            });
+        }
+
+        export_function(name, fn) {
+            this.function_table[name] = fn;
+        }
+
+        evaljs(code, replacements) {
```

Comment:
Doesn't evaljs need to be sync? e.g. in `JSContext.displatch_event` ?

Reply:
Ah—our compiler compiles everything to async, so `evaljs` is sync in Python but that becomes async in JS.

Reply:
I see. Does that mean that preventDefault doesn't work in the widgets?

Reply:
Which `preventDefault` do you mean? When the user clicks in the Canvas, that generates an event, which doesn't have a default action. When we send that to the browser, all of the calls use `await f()` and `async function ...`, but it works the same way. I think `preventDefault` should work; that's just something returned to the caller of `evaljs` and then to the caller of `dispatch_event` which then knows whether to do the default action.

Here's a way to think about it. Imagine a subset of JS where every function call is written `await f()` and every function definition is written `async function`... That's the subset we compile Python to. (There are a few workarounds for parts of JS where this doesn't work.)

Reply:
ah right. thanks for explaining.